### PR TITLE
Add get_lienar_qubit_operator_diagonal to be used for Davidson algorithm.

### PR DESCRIPTION
**Pull request by sheilaliuxl**
This's to get the diagonal elements for the matrix only, which is a critical input for Davidson algorithm.

See one reference at:
http://people.inf.ethz.ch/arbenz/ewp/Lnotes/chapter12.pdf

### CODE REVIEW

File: src/openfermion/utils/_sparse_tools.py [hunk 1]
```diff
@@ -283,6 +283,58 @@ def matvec(vec):
...
+            # Split vector by half and half for each bit.
+            if pauli_operator[0] > tensor_factor:
+                vecs = [v for iter_v in vecs for v in numpy.split(
+                    iter_v, 2 ** (pauli_operator[0] - tensor_factor))]
+
+            vec_pairs = [numpy.split(v, 2) for v in vecs]
+            vecs = [v for vp in vec_pairs for v in (vp[0], -vp[1])]
+            tensor_factor = pauli_operator[0] + 1
+        if not is_zero:
+            linear_operator_diagonal += qubit_operator.terms[qubit_term] * \
```
Comment (jarrodmcc) [inline src/openfermion/utils/_sparse_tools.py]: Just a tiny style thing, but in the codebase we prefer using
x = ( something
    new line someting)
to 
x = something \
   new line something

File: src/openfermion/utils/_sparse_tools_test.py [hunk 2]
```diff
@@ -183,6 +193,61 @@ def test_get_linear_qubit_operator_compare(self):
...
+            get_linear_qubit_operator_diagonal(qubit_operator), vec_expected))
+
+    def test_get_linear_qubit_operator_diagonal_cmp_zero(self):
+        """Compare get_linear_qubit_operator_diagonal with
+            get_linear_qubit_operator.
+        """
+        qubit_operator = QubitOperator('Z1 X2 Y5')
+
+        vec = get_linear_qubit_operator_diagonal(qubit_operator)
+        vec_expected = numpy.diag(get_linear_qubit_operator(qubit_operator) * \
```
Comment (jarrodmcc) [inline src/openfermion/utils/_sparse_tools_test.py]: newline comment from above

File: src/openfermion/utils/_sparse_tools_test.py [hunk 3]
```diff
@@ -183,6 +193,61 @@ def test_get_linear_qubit_operator_compare(self):
...
+            get_linear_qubit_operator_diagonal(qubit_operator), vec_expected))
+
+    def test_get_linear_qubit_operator_diagonal_cmp_non_zero(self):
+        """Compare get_linear_qubit_operator_diagonal with
+            get_linear_qubit_operator.
+        """
+        qubit_operator = QubitOperator('Z1 Z2 Z5')
+
+        vec = get_linear_qubit_operator_diagonal(qubit_operator)
+        vec_expected = numpy.diag(get_linear_qubit_operator(qubit_operator) * \
```
Comment (jarrodmcc) [inline src/openfermion/utils/_sparse_tools_test.py]: same

File: src/openfermion/utils/_sparse_tools.py (hunk 1 above)
Comment (sheilaliuxl) [reply src/openfermion/utils/_sparse_tools.py]: Got it, thanks for the suggestion!

File: src/openfermion/utils/_sparse_tools_test.py (hunk 2 above)
Comment (sheilaliuxl) [reply src/openfermion/utils/_sparse_tools_test.py]: Fixed.

File: src/openfermion/utils/_sparse_tools_test.py (hunk 3 above)
Comment (sheilaliuxl) [reply src/openfermion/utils/_sparse_tools_test.py]: Fixed too.

**Pull request merged**